C++: Clean up BufferWrite.qll

geoffw0 opened the pull request:
While working on the QLDoc for BufferWrite.qll, I noticed some places where it could depend more on models (reducing redundancy), and where models can be improved using cases from BufferWrite.qll.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/models/implementations/Strcat.qll:
```diff
@@ -24,6 +24,21 @@ class StrcatFunction extends TaintFunction, DataFlowFunction, ArrayFunction, Sid
+  /**
+   * Gets the index of the parameter that is the size of the copy (in characters).
+   */
+  int getParamSize() { if exists(getParameter(2)) then result = 2 else none() }
```
Writing this as `exists(getParameter(2)) and result = 2` feels more QL'y to me.

geoffw0 replied:
Agreed.  Updated.

MathiasVP commented:
Other than the failing ql{doc, format} checks, I think this LGTM! The new test result also looks right to me.

geoffw0 commented:
Fixed formatting and fixed the new false positive in `UsingStrcpyAsBoolean.ql`.  It was caused by the query assuming that all `strcpy` variants return a string, whereas some of the new ones (e.g. `strcpy_s`) do in fact return an error code so the query doesn't apply to them.
